### PR TITLE
Feature/reviews and admin page logic

### PR DESCRIPTION
**Pull request by Ahontic**
Step 6
Reviews on Book page and Admin page logic

### CODE REVIEW

File: app/controllers/addresses_controller.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 class AddressesController < ApplicationController
   def create
-    @address = current_customer.addresses.new(address_params)
+    @address = current_customer.addresses.create(address_params)
```
Comment (6aBblKuH) [inline app/controllers/addresses_controller.rb]: move it back

File: app/controllers/reviews_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class ReviewsController < ApplicationController
+  decorates_assigned :book
+
+  def index
+    @book = BookDecorator.find_by(id: params[:book_id])
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: `book = Book.find(params[:book_id])`

File: app/controllers/reviews_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    redirect_to category_book_path(id: @book.id, category_id: @book.category_id)
+  end
+
+  def create
+    @book = BookDecorator.find_by(id: params[:book_id])
+    @review = Review.new(review_params)
+    @reviews = @book.reviews.approved
+    if @review.save
+      flash[:notice] = 'Thanks for Review. It will be published as soon as Admin will approve it.'
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: i18n

File: app/controllers/reviews_controller.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  decorates_assigned :book
+
+  def index
+    @book = BookDecorator.find_by(id: params[:book_id])
+    redirect_to category_book_path(id: @book.id, category_id: @book.category_id)
+  end
+
+  def create
+    @book = BookDecorator.find_by(id: params[:book_id])
+    @review = Review.new(review_params)
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]:
```ruby
@book.reviews.new(customer: current_customer)
```

File: app/controllers/reviews_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    @reviews = @book.reviews.approved
+    if @review.save
+      flash[:notice] = 'Thanks for Review. It will be published as soon as Admin will approve it.'
+      redirect_back(fallback_location: root_path)
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: ```redirect_to category_book_path(@book.category_id, @book.id)```
or just render anyway book/show

File: app/controllers/reviews_controller.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    else
+      render 'books/show'
+    end
+  end
+
+  private
+
+  def review_params
+    params.require(:review).permit(:title, :body, :rating, :book_id, :customer_id)
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: exclude from params `:book_id, :customer_id`

File: app/views/reviews/_form.html.haml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+= simple_form_for([@book, @book.reviews.build]) do |f|
+  #rating-form
+    %label Rating
+  .form-group
+    = f.input :book_id, label: false, input_html: { value: @book.id, type: 'hidden' }
+    = f.input :customer_id, label: false, input_html: { value: current_customer.id, type: 'hidden' }
```
Comment (6aBblKuH) [inline app/views/reviews/_form.html.haml]: remove these fields

File: app/views/reviews/_review.html.haml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+- @reviews.each do |review|
+  - if "#{@reviews.count}" > "0"
```
Comment (6aBblKuH) [inline app/views/reviews/_review.html.haml]: `@reviews.any?`

File: db/migrate/20200103150503_add_customer_id_to_reviews.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddCustomerIdToReviews < ActiveRecord::Migration[6.0]
+  def change
+    add_column :reviews, :customer_id, :integer
```
Comment (6aBblKuH) [inline db/migrate/20200103150503_add_customer_id_to_reviews.rb]: it should be foreign key
move it to main migration pleaase

File: db/migrate/20200103150604_add_book_id_to_reviews.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddBookIdToReviews < ActiveRecord::Migration[6.0]
+  def change
+    add_column :reviews, :book_id, :integer
```
Comment (6aBblKuH) [inline db/migrate/20200103150604_add_book_id_to_reviews.rb]: it should be foreign key
move it to main migration pleaase

File: db/migrate/20200104090845_add_status_to_reviews.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddStatusToReviews < ActiveRecord::Migration[6.0]
+  def change
+    add_column :reviews, :status, :integer, default: 0
```
Comment (6aBblKuH) [inline db/migrate/20200104090845_add_status_to_reviews.rb]: move it to main migration pleaase

File: spec/models/review_spec.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+#  status      :integer          default("pending")
+#  title       :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#  book_id     :integer
+#  customer_id :integer
+#
+
+RSpec.describe Review do
+  let(:review) { create(:review) }
```
Comment (6aBblKuH) [inline spec/models/review_spec.rb]: remove it

File: spec/models/review_spec.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+  describe 'validations' do
+    it 'is expected to validate that :status is pending' do
```
Comment (6aBblKuH) [inline spec/models/review_spec.rb]: ```is_expected.to define_enum_for(:name).with_values(:a, :b)```

File: spec/features/review_spec.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        fill_in 'Review', with: 'Review test'
+      end
+      click_button 'Post'
+      expect(page).to have_content 'Thanks for Review'
+
+      visit category_book_path(category_id: book.category_id, id: book.id)
+      expect(page).to_not have_content 'Review test'
+      review = Review.last
+      review.status = 'approved'
+      review.save
```
Comment (6aBblKuH) [inline spec/features/review_spec.rb]: `book.reviews.first.approved!`

File: spec/features/review_spec.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      page.driver.browser.navigate.refresh
```
Comment (6aBblKuH) [inline spec/features/review_spec.rb]: `visit category_book_path book.category_id, book.id`

File: spec/features/review_spec.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+feature 'Book page' do
+  let!(:customer) { create(:customer, :confirmed) }
+  let!(:book) { create(:book, :with_author) }
+  let!(:review_approved) { create(:review, book: book, customer: customer, status: 'approved') }
+  let!(:review_pending) { create(:review, book: book, customer: customer) }
+  let!(:review_rejected) { create(:review, book: book, customer: customer, status: 'rejected') }
```
Comment (6aBblKuH) [inline spec/features/review_spec.rb]: move it to concrete spec case

File: app/controllers/application_controller.rb [hunk 17]
```diff
@@ -1,10 +1,19 @@
...
   def not_found
     render file: Rails.public_path.join('404.html'), status: :not_found, layout: false
   end
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
+    devise_parameter_sanitizer.permit(:account_update, keys: %i[name avatar])
+  end
```
Comment (6aBblKuH) [inline app/controllers/application_controller.rb]: use instead image from oauth response

File: Gemfile [hunk 18]
```diff
@@ -14,6 +14,7 @@ gem 'draper'
 gem 'figaro'
 gem 'haml-rails'
 gem 'i18n'
+gem 'image_processing', '~> 1.2'
```
Comment (6aBblKuH) [inline Gemfile]: how do u use  it?

File: app/controllers/reviews_controller.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+  def index
+    book = Book.find(params[:book_id])
+    redirect_to category_book_path(book.category_id, book.id)
+  end
+
+  def create
+    @book = BookDecorator.find_by(id: params[:book_id])
+    @review = @book.reviews.new(review_params.merge(customer: current_customer))
+    @reviews = @book.reviews.approved
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: remove

File: app/controllers/reviews_controller.rb [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+class ReviewsController < ApplicationController
+  decorates_assigned :book
+
+  def index
+    book = Book.find(params[:book_id])
+    redirect_to category_book_path(book.category_id, book.id)
+  end
+
+  def create
+    @book = BookDecorator.find_by(id: params[:book_id])
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: @book = Book.find(params[:book_id])

File: app/controllers/reviews_controller.rb [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+    @review = @book.reviews.new(review_params.merge(customer: current_customer))
```
Comment (6aBblKuH) [inline app/controllers/reviews_controller.rb]: make it local var

File: app/views/reviews/_review.html.haml [hunk 22]
```diff
@@ -0,0 +1,16 @@
+- @reviews.each do |review|
+  - if @reviews.any?
```
Comment (6aBblKuH) [inline app/views/reviews/_review.html.haml]: remove

File: db/migrate/20200103144825_create_reviews.rb [hunk 23]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class CreateReviews < ActiveRecord::Migration[6.0]
+  def change
+    create_table :reviews do |t|
```
[the remark this answers is not in the document]
Comment (6aBblKuH) [reply db/migrate/20200103144825_create_reviews.rb]: null false

File: spec/models/review_spec.rb [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+#  fk_rails_...  (book_id => books.id)
+#  fk_rails_...  (customer_id => customers.id)
+#
+
+RSpec.describe Review do
+  describe 'validations' do
+    it { is_expected.to validate_presence_of(:body) }
+    it { is_expected.to validate_presence_of(:rating) }
+    it { is_expected.to validate_presence_of(:title) }
+    it { should define_enum_for(:status).with_values(pending: 0, approved: 1, rejected: 2) }
```
Comment (6aBblKuH) [inline spec/models/review_spec.rb]: is_expected.to

File: app/views/reviews/_review.html.haml (hunk 22 above)
Comment (Ahontic) [reply app/views/reviews/_review.html.haml]: removed

File: app/controllers/reviews_controller.rb (hunk 20 above)
Comment (Ahontic) [reply app/controllers/reviews_controller.rb]: done

File: app/controllers/reviews_controller.rb (hunk 19 above)
Comment (Ahontic) [reply app/controllers/reviews_controller.rb]: done

File: app/controllers/reviews_controller.rb (hunk 21 above)
Comment (Ahontic) [reply app/controllers/reviews_controller.rb]: done

File: db/migrate/20200103144825_create_reviews.rb (hunk 23 above)
[the remark this answers is not in the document]
Comment (Ahontic) [reply db/migrate/20200103144825_create_reviews.rb]: done

File: spec/models/review_spec.rb (hunk 24 above)
Comment (Ahontic) [reply spec/models/review_spec.rb]: done

**Pull request merged**